### DISCUSSION TITLE
function properties problem

### DISCUSSION THREAD

**Discussion by ret2void** [Vector35/binaryninja-api#4343] [Q&A] [answered]
![image](https://github.com/Vector35/binaryninja-api/assets/83763549/89eccc10-2359-4c55-93a2-cda1f5d91cfd)
The 'chart_course' function is supposed to take only one argument, arg1, but the parsing results in a second argument, zmm0. Now I Edit the Function with 'Edit Function Properties' :
![image](https://github.com/Vector35/binaryninja-api/assets/83763549/e286e74b-ce2d-4990-be85-426a9da0063b)
I unchecked zmm0,but it still appears：
![image](https://github.com/Vector35/binaryninja-api/assets/83763549/8238d8e9-9bf4-4937-b80b-ac959ade2dca)
another question is function 'atof’,:
![image](https://github.com/Vector35/binaryninja-api/assets/83763549/e0851ff2-5a61-4af6-9b41-09bf4bb753d8)
as you see, the return register of this function should be xmm0，and I also used the same way to edit function properties,but nothing happen.
![image](https://github.com/Vector35/binaryninja-api/assets/83763549/d5a3b807-4526-4e5e-aaab-55169c85ac38)
in next picture, the arg2 should be replaced by var_78.
![image](https://github.com/Vector35/binaryninja-api/assets/83763549/2ca7555e-59c8-4240-9b9b-5c7ad96be307)

is it a bug?
the binary file is here:
https://github.com/guyinatuxedo/nightmare/blob/master/modules/08-bof_dynamic/fb19_overfloat/overfloat

**Comment by psifertex** [comment]
Yeah, this is definitely a problem and while you can fix the parent function having the extra zmm0 parameter, it won't ultimately fix the decompilation being wrong with that return function. I'm scheduling this one for resolution for the next stable and we'll be working on it but I don't have a specific timeline for a fix or unfortunately even a way to manually work-around it at this point.

**Reply by psifertex** [reply, marked as answer]
I'm marking this as the answer -- the first part of the question above is actually a misunderstanding of what clobbers do. Clobbers don't control arguments at all, they're related to the calling convention. The way to fix an incorrect number of arguments is to change the type either with `y` if it's always a single argument, or right-click and choose "override call type" if it's a variadic function for example like printf where the argument count can change.

As to the second issue, that's actually #4027 and is indeed an issue that we're tracking there.

**Reply by ret2void** [reply]
override call type seems can not set the returned value's register. Is there a way?

**Reply by psifertex** [reply]
Sorry for not replying to this one earlier. This requires a custom calling convention as the return register is specifically part o the CC, not the type information.

**Comment by ret2void** [comment]
I have found a new bug which occured in debugger:
OS:Ubuntu 22.04 lts
VERSION:3.5.4377
breakpoint may be invalid!
I used the debugger to debug the attachments I uploaded, but I found that any breakpoints seemed invalid, even down to the first instruction of the program.

First, I toggle a breakpoint as follows:

Then I just start debugger and click "Run"，but breakpoint is not breaked!

------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Vector35/binaryninja-api"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年6月1日(星期四) 凌晨2:48
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Vector35/binaryninja-api] function properties problem (Discussion #4343)





 
Yeah, this is definitely a problem and while you can fix the parent function having the extra zmm0 parameter, it won't ultimately fix the decompilation being wrong with that return function. I'm scheduling this one for resolution for the next stable and we'll be working on it but I don't have a specific timeline for a fix or unfortunately even a way to manually work-around it at this point.

**Reply by psifertex** [reply]
Please file debugger issues over at: https://github.com/vector35/debugger/issues/new
